moby/moby
# add dockerd-rootless-setuptool.sh

**- What I did**

Added `contrib/dockerd-rootless-setuptool.sh`.

The tool is expected to be used by https://get.docker.com/rootless script.

Can be included in RPM/DPKG packages as well.

Fix https://github.com/moby/moby/issues/40707
Relates to https://github.com/moby/moby/issues/40761
Replaces https://github.com/moby/moby/pull/40924 (Same as this PR but written in Go rather than in shell)

**- How I did it**
Rewrote https://get.docker.com/rootless .
The curl stuff is expected to be kept in https://get.docker.com/rootless .

Also, the storage driver is not detected in the new tool because the daemon should be able to choose overlay2/fuse-overlayfs/vfs automatically.

**- How to verify it**

Show usage:
```console
$ dockerd-rootless-setuptool.sh --help
Usage: dockerd-rootless-setuptool.sh [OPTIONS] COMMAND

A setup tool for Rootless Docker (dockerd-rootless.sh).

Documentation: https://docs.docker.com/engine/security/rootless/

Options:
  -f, --force                Ignore rootful Docker (/var/run/docker.sock)
      --skip-iptables        Ignore missing iptables

Commands:
  check        Check prerequisites
  install      Install systemd unit (if systemd is available) and show how to manage the service
  uninstall    Uninstall systemd unit
```

Install (-> fail due to existing rootful Docker)
```console
$ dockerd-rootless-setuptool.sh install
[ERROR] Aborting because rootful Docker (/var/run/docker.sock) is running and accessible. Set --force to ignore.
```

Install (ignore rootful Docker -> fail due to misconfiguration of subuid and subgid)
```console
$ dockerd-rootless-setuptool.sh install --force
[ERROR] Missing system requirements. Run the following commands to
[ERROR] install the requirements and run this tool again.

########## BEGIN ##########
cat <<EOF | sudo sh -x
# Add subuid entry for suda
echo "suda:100000:65536" >> /etc/subuid
# Add subgid entry for suda
echo "suda:100000:65536" >> /etc/subgid
EOF
########## END ##########
```

Install (try again after fixing up subuid and subgid -> success)
```console
$ cat <<EOF | sudo sh -x
> # Add subuid entry for suda
> echo "suda:100000:65536" >> /etc/subuid
> # Add subgid entry for suda
> echo "suda:100000:65536" >> /etc/subgid
> EOF
+ echo suda:100000:65536
+ echo suda:100000:65536
$ dockerd-rootless-setuptool.sh install --force
[INFO] Creating /home/suda/.config/systemd/user/docker.service
[INFO] starting systemd service docker.service
+ systemctl --user start docker.service
+ sleep 3
+ set +x
+ systemctl --user --no-pager --full status docker.service
● docker.service - Docker Application Container Engine (Rootless)
     Loaded: loaded (/home/suda/.config/systemd/user/docker.service; disabled; vendor preset: enabled)
     Active: active (running) since Mon 2020-05-11 23:37:26 JST; 3s ago
       Docs: https://docs.docker.com/engine/security/rootless/
   Main PID: 15810 (rootlesskit)
      Tasks: 38
     Memory: 27.3M
        CPU: 397ms
     CGroup: /user.slice/user-1001.slice/user@1001.service/docker.service
             ├─15810 rootlesskit --net=slirp4netns --mtu=65520 --slirp4netns-sandbox=auto --slirp4netns-seccomp=auto --disable-host-loopback --port-driver=builtin --copy-up=/etc --copy-up=/run --propagation=rslave /usr/local/bin/dockerd-rootless.sh
             ├─15836 /proc/self/exe --net=slirp4netns --mtu=65520 --slirp4netns-sandbox=auto --slirp4netns-seccomp=auto --disable-host-loopback --port-driver=builtin --copy-up=/etc --copy-up=/run --propagation=rslave /usr/local/bin/dockerd-rootless.sh
             ├─15852 slirp4netns --mtu 65520 -r 3 --disable-host-loopback --enable-sandbox --enable-seccomp 15836 tap0
             ├─15860 dockerd
             └─15877 containerd --config /run/user/1001/docker/containerd/containerd.toml --log-level info
...
+ DOCKER_HOST=unix:///run/user/1001/docker.sock /usr/local/bin/docker version
Client:
 Version:           20.03.0-dev
 API version:       1.41
 Go version:        go1.13.10
 Git commit:        f7185d27e
 Built:             Fri May  8 05:21:12 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          dev
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       cdf49f37ee
  Built:            Fri May  8 05:18:12 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.0-551-g990076b7
  GitCommit:        990076b731ec9446437972b41176a6b0f3b7bcbf
 runc:
  Version:          1.0.0-rc10+dev
  GitCommit:        2b31437caa905b7b944a891aee613e7dd0a1f898
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
+ systemctl --user enable docker.service
Created symlink /home/suda/.config/systemd/user/default.target.wants/docker.service → /home/suda/.config/systemd/user/docker.service.
+ set +x
[INFO] Installed docker.service successfully.
[INFO] To control docker.service, run: `systemctl --user (start|stop|restart) docker.service`
[INFO] To run docker.service on system startup, run: `sudo loginctl enable-linger suda`

[INFO] Make sure the following environment variables are set (or add them to ~/.bashrc):

export PATH=/usr/local/bin:$PATH
export DOCKER_HOST=unix:///run/user/1001/docker.sock
```

Uninstall:
```console
$ dockerd-rootless-setuptool.sh uninstall --force
+ systemctl --user stop docker.service
+ systemctl --user disable docker.service
Removed /home/suda/.config/systemd/user/default.target.wants/docker.service.
+ rm -f /home/suda/.config/systemd/user/docker.service
+ set +x -e
[INFO] Uninstalled docker.service
[INFO] This uninstallation tool does NOT remove Docker binaries and data.
[INFO] To remove data, run: `/usr/local/bin/rootlesskit rm -rf /home/suda/.local/share/docker`
```


**- Description for the changelog**

add dockerd-rootless-setuptool.sh

**- A picture of a cute animal (not mandatory but encouraged)**

:penguin:
